File: environments/aws.go
package environments

type AwsAuthenticationConfig struct {
	Profile    *string `mapstructure:"profile"`
	AssumeRole *string `mapstructure:"assume_role"`
	Account    string  `mapstructure:"account"`
	Region     string  `mapstructure:"region"`
}

func (aws *AwsAuthenticationConfig) EnvVars() map[string]string {
	envVars := map[string]string{}

	envVars["AWS_REGION"] = aws.Region
	envVars["AWS_DEFAULT_REGION"] = aws.Region
	envVars["AWS_ACCOUNT_ID"] = aws.Account

	if aws.Profile != nil {
		envVars["AWS_PROFILE"] = *aws.Profile
		envVars["AWS_DEFAULT_PROFILE"] = *aws.Profile
	}

	return envVars
}

func (dest *AwsAuthenticationConfig) Merge(src *AwsAuthenticationConfig) {
	if src.Profile != nil {
		dest.Profile = src.Profile
	}
	if src.AssumeRole != nil {
		dest.AssumeRole = src.AssumeRole
	}
	dest.Account = src.Account
	dest.Region = src.Region
}

File: src/environments/aws.ts
import { EnvironmentConfig } from './index'

export class AwsAuthenticationConfig implements EnvironmentConfig<AwsAuthenticationConfig> {
  profile: string | undefined
  assumeRole: string | undefined
  account: string = ""
  region: string = ""

  constructor(config: {
    account: string,
    region: string,
    profile: string | undefined,
    assumeRole: string | undefined
  }) {
    Object.assign(this, config)
  }

  env(): { [key: string]: string } {
    var env: { [key: string]: string } = {
      AWS_REGION: this.region,
      AWS_DEFAULT_REGION: this.region,
      AWS_ACCOUNT_ID: this.account
    }

    if (this.profile !== undefined) {
      env["AWS_PROFILE"] = this.profile
      env["AWS_DEFAULT_PROFILE"] = this.profile
    }

    return env
  }

  merge(...envs: AwsAuthenticationConfig[]): void {
    for (let env of envs) {
      this.account = env.account
      this.region = env.region

      if (env.profile !== undefined) {
        this.profile = env.profile
      }

      if (env.assumeRole !== undefined) {
        this.assumeRole = env.assumeRole
      }
    }
  }
}
